### PR TITLE
[Preview] React Component pane

### PR DESCRIPTION
**Pull request by mmcote**
Fixes Issue: #3800 -- 5: Show component instances in the sidebar.

### Summary of Changes

* Added new FrameworkComponent.
* Separated out function to get extra props.

### Test Plan

* Currently validated through the browser, will add tests.

### Final Screenshot
screen shot 2018-03-27 at 10 50 29 am

### CODE REVIEW

**Comment by jasonLaster** [comment]
Awesome!

where is `t` coming from. It looks like the component is `App`

**Comment by mmcote** [comment]
@jasonLaster, yeah for sure, I was curious about this as well, currently I am just utilizing what was there for the popup, which returns the same name, as they are both using the `getReactProps`. Which 
essentially is just calling `this._reactInternalFiber.type.name` or `this._reactInternalInstance.getName()` depending on the react version. I will look into each react object and find a way to return the proper name. 

screen shot 2018-03-27 at 11 18 01 am

**Comment by mmcote** [comment]
^^^ Change that, I am not looking for a new place to get the name, just looking for why I'm currently getting t, and what needs to be changed for these statements to return the correct name.

**Comment by jasonLaster** [comment]
sounds good

**Comment by mmcote** [comment]
@jasonLaster, the reason the `t` was coming appearing was because the react applications I was testing this feature on were deployed react apps that have been minified for production. 

This was stated here, as they had the same problem: [https://github.com/facebook/react-native/issues/10550](https://github.com/facebook/react-native/issues/10550), and repeated here in the last comment: [https://stackoverflow.com/questions/43800784/get-component-name-in-react](https://stackoverflow.com/questions/43800784/get-component-name-in-react).

Tested this myself:
Development
screen shot 2018-03-27 at 12 35 10 pm

Production
screen shot 2018-03-27 at 12 33 52 pm

I will look if there is a way of getting the name in production.

**Comment by jasonLaster** [comment]
I see, that's not a big deal. We have one other option too you know :)

If we know that, 

a) `this` is a react component, 
b) that the debug line is in a class we can use the symbols to get the closest class
c) that the debug line is not in a class, we can use the symbols table to get the closest function.

**Comment by mmcote** [comment]
Oh awesome, okay I will look at implementing this as an alternative when not in a dev environment. :)

**Comment by jasonLaster** [comment]
Sounds good, i'll see if we can merge this as is for now

File: src/actions/preview.js [hunk 1]
```diff
@@ -90,6 +90,20 @@ function isInvalidTarget(target: HTMLElement) {
   return invalidTarget || invalidToken || invaildType;
 }
 
+export function getExtra(
+  expression: string,
+  result: Object,
+  selectedFrame: any
```
Comment (jasonLaster) [inline src/actions/preview.js]: `selectedFrame: Frame`

File: src/components/SecondaryPanes/FrameworkComponent.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+import { ObjectInspector, ObjectInspectorUtils } from "devtools-reps";
+import { isReactComponent } from "../../utils/preview";
+
+const { createNode, getChildren } = ObjectInspectorUtils.node;
+const { loadItemProperties } = ObjectInspectorUtils.loadProperties;
+
+type Props = {
+  setPopupObjectProperties: (Object, Object) => void,
+  frame: any,
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/FrameworkComponent.js]: `frame: Frame`

File: src/components/SecondaryPanes/index.js [hunk 3]
```diff
@@ -151,6 +154,20 @@ class SecondaryPanes extends Component<Props> {
     };
   }
 
+  getComponentItem() {
+    const { extra: { react } } = this.props;
+
+    return {
+      header: react.displayName,
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/index.js]: could this be null?

File: src/components/SecondaryPanes/index.js [hunk 4]
```diff
@@ -241,6 +258,10 @@ class SecondaryPanes extends Component<Props> {
       items.push(this.getCallStackItem());
       if (this.props.horizontal) {
         items.push(this.getScopeItem());
+
+        if (extra.react) {
```
Comment (jasonLaster) [inline src/components/SecondaryPanes/index.js]: guess not :)

File: src/utils/prefs.js [hunk 5]
```diff
@@ -19,6 +19,7 @@ if (isDevelopment()) {
   pref("devtools.debugger.ignore-caught-exceptions", false);
   pref("devtools.debugger.call-stack-visible", true);
   pref("devtools.debugger.scopes-visible", true);
+  pref("devtools.debugger.component-visible", true);
```
Comment (jasonLaster) [inline src/utils/prefs.js]: we'll also have to add this to assets/prefs

**Pull request merged**